"borrow checker invariants" section of the (#2867)
google/comprehensive-rust: src/idiomatic/leveraging-the-type-system/borrow-checker-invariants/generalizing-ownership.md

---
minutes: 0

> ---

Reviewer: Missing the title on this one.

The logic of the borrow checker, while modelled off "memory ownership", can be abstracted away from that use case to model other problems where we want to prevent API misuse.

```rust,editable
pub struct InternalData;
pub struct Value(InternalData);

fn shared_use(value: &Value) -> &InternalData {
    &value.0
}

fn exclusive_use(value: &mut Value) -> &mut InternalData {
    &mut value.0
}

fn deny_future_use(value: Value) {}

let mut value = Value(InternalData);
let deny_mut = shared_use(&value);
let try_to_deny_immutable = exclusive_use(&mut value); // ❌🔨
let more_mut_denial = &deny_mut;
deny_future_use(value);

> let even_more_mut_denial = shared_use(&value); // ❌🔨

Reviewer: In the speaker notes, we should elaborate as to what the example is trying to show and it relates to the concepts we're discussing.
This is a bit too implicit for me on this slide.

```

<details>

- To use the borrow checker as a problem solving tool, we will need to "forget" that the original purpose of it is to prevent mutable aliasing in the context of concurrency, instead imagining and working within situations where the rules are the same but the meaning is slightly different.

- In rust's borrow checker we have access to three different ways of "taking" a value:

> <!-- TODO: actually link to the RAII section when it has been merged. -->

Author: The important part here is the context – the rules are being laid out in with the knowledge that we're going to be using them to restrict the shape of APIs rather than specifically to avoid memory issues. You're right that it could be more clear that this is the context these rules are being reiterated in :)

---
minutes: 5
---

# Generalizing Ownership: Lifetimes Refresher

The logic of the borrow checker, while modelled off "memory ownership", can be
abstracted away from that use case to model other problems where we want to
prevent API misuse.

```rust,editable
// An internal data type to have something to hold onto.
pub struct Internal;
// The "outer" data.
pub struct Data(Internal);

fn shared_use(value: &Data) -> &Internal {
    &value.0
}
fn exclusive_use(value: &mut Data) -> &mut Internal {
    &mut value.0
}
fn deny_future_use(value: Data) {}

fn main() {
    let mut value = Data(Internal);
    let shared = shared_use(&value);
    // let exclusive = exclusive_use(&mut value); // ❌🔨
    let shared_again = &shared;
    deny_future_use(value);

>     // let shared_again_again = shared_use(&value); // ❌🔨

Reviewer 2: I find this code example more clear than before, thank you!

I do feel though that 3 ideas (what each parameter passing mode achieves) are too cramped in one function and interfere with each other. Could we split this into 3 pseudo-main functions? It does not matter that the code would not have an actual main, there is nothing runnable here anyway.

---
minutes: 0
---

# Generalizing Ownership

The logic of the borrow checker, while modelled off "memory ownership", can be
abstracted away from that use case to model other problems where we want to
prevent API misuse.

```rust,editable,compile_fail
// An internal data type to have something to hold onto.
pub struct Internal;
// The "outer" data.
pub struct Data(Internal);

fn shared_use(value: &Data) -> &Internal {
    &value.0
}
fn exclusive_use(value: &mut Data) -> &mut Internal {
    &mut value.0
}
fn deny_future_use(value: Data) {}

fn main() {
    let mut value = Data(Internal);
    let deny_mut = shared_use(&value);
    let try_to_deny_immutable = exclusive_use(&mut value); // ❌🔨
    let more_mut_denial = &deny_mut;
    deny_future_use(value);
    let even_more_mut_denial = shared_use(&value); // ❌🔨
}
```

<details>

- This example re-frames the borrow checker rules away from references and
  towards semantic meaning in non-memory-safety settings. Nothing is being
  mutated, nothing is being sent across threads.

- To use the borrow checker as a problem solving tool, we will need to "forget"
  that the original purpose of it is to prevent mutable aliasing in the context

>   of concurrency & dangling pointers, instead imagining and working within

Reviewer 2: ```suggestion
  of preventing use-after-frees and data races, instead imagining and working within
```

---
minutes: 10
---

# Lifetimes and Borrows: the Abstract Rules

Refresher on borrow checker basics.

> 

Reviewer 2: ```suggestion
```

I think the title is sufficiently strong, this sentence does not add much. Let's have more vertical space on the slide instead.

  towards semantic meaning in non-memory-safety settings.

  Nothing is being mutated, nothing is being sent across threads.

- In rust's borrow checker we have access to three different ways of "taking" a
  value:

  - Owned value `T`. Value is dropped when the scope ends, unless it is not
    returned to another scope.

  - Shared Reference `&T`. Allows aliasing but prevents mutable access while
    shared references are in use.

  - Mutable Reference `&mut T`. Only one of these is allowed to exist for a
    value at any one point, but can be used to create shared references.

- Ask: The two commented-out lines in the `demo` functions would cause
  compilation errors, Why?

  `demo_exclusive`: Because the `shared` value is still aliased after the
  `exclusive` reference is taken.

  `demo_denied`: Because `value` is consumed (AKA dropped) the line before the
  `shared_again_again` reference is taken from `&value`.

- Remember that every `&T` and `&mut T` has a lifetime, just one the user
  doesn't have to annotate or think about most of the time.

  We get to avoid annotating a lot of lifetimes because the rust compiler allows
  a user to elide the majority of them. See:

>   [Lifetime Elision](../../../lifetimes/lifetime-elision.md)

Reviewer 2: ```suggestion
  We rarely specify lifetimes because the Rust compiler allows
  us to *elide* them in most cases. See:
  [Lifetime Elision](../../../lifetimes/lifetime-elision.md)
```

Simpler phrasing, italics on "elide" to emphasize a technical term.

---
minutes: 15
---

# Generalizing Ownership

The logic of the borrow checker, while modelled off "memory ownership", can be
abstracted away from that use case to model other problems where we want to
prevent API misuse.

```rust,editable,compile_fail
// An internal data type to have something to hold onto.
pub struct Internal;
// The "outer" data.
pub struct Data(Internal);

fn shared_use(value: &Data) -> &Internal {
    &value.0
}
fn exclusive_use(value: &mut Data) -> &mut Internal {
    &mut value.0
}
fn deny_future_use(value: Data) {}

fn main() {
    let mut value = Data(Internal);
    let deny_mut = shared_use(&value);

>     let try_to_deny_immutable = exclusive_use(&mut value); // ❌🔨

Reviewer 2: If this line does not compile because of the function call, what does the variable declaration add? Why not let it simply be `exclusive_use(&mut value);`?

Similarly for the line below that also does not compile.

    let more_mut_denial = &deny_mut;
    deny_future_use(value);
    let even_more_mut_denial = shared_use(&value); // ❌🔨

> }

Reviewer 2: The code is a good demonstration of the borrow checker rules, but it doesn't connect to a real-world problem.

Is it possible to find a bit more engaging example to illustrate the point? For example, make a simplified implementation of `BorrowedFd` and illustrate the same points with that.

Author: Thinking about it more, a large chunk of this slide can be moved to the introductory slide for this chapter and the rest can be re-purposed as a brief borrowck refresher.
